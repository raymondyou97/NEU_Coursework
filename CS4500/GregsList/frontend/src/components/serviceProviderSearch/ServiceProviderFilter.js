import React from 'react';

class ServiceProviderFilter extends React.Component {
  constructor(props) {
    super(props);

    this.newPredicates = this.props.searchPredicates;
  }

  displayFilters() {
    let searchPredicates = this.props.searchPredicates;
    let questionCount = 0;
    let finalDisplay = searchPredicates.map(predicate => {
      const question = predicate.serviceQuestion;
      questionCount++;
      let choices = question.choices.split(',');
      switch (question.type) {
        case 'MultipleChoice':
          return (
            <div key={question.id} className="service-filter-question">
              <h5>
                Question {questionCount}: {question.question}
              </h5>
              {this.getMultipleChoice(questionCount, choices, i => {
                predicate.serviceAnswer.choiceAnswer = i;
                this.newPredicates = searchPredicates;
              })}
            </div>
          );
        case 'TrueFalse':
          return (
            <div key={question.id} className="service-filter-question">
              <h5>
                Question {questionCount}: {question.question}
              </h5>
              {this.getTrueFalse(questionCount, choices, bool => {
                predicate.serviceAnswer.trueFalseAnswer = bool;
                this.newPredicates = searchPredicates;
              })}
            </div>
          );
        case 'Range':
          return (
            <div key={question.id} className="service-filter-question">
              <h5>
                Question {questionCount}: {question.question}
              </h5>
              {this.getRange(questionCount, choices, rangeVal => {
                predicate.serviceAnswer.minRangeAnswer = rangeVal;
                predicate.serviceAnswer.maxRangeAnswer = rangeVal;
                this.newPredicates = searchPredicates;
              })}
            </div>
          );
        default:
          break;
      }
      return [];
    });

    return finalDisplay;
  }

  getMultipleChoice(questionCount, choices, handleChange) {
    let finalView = [];
    for (let i = 0; i < choices.length; i++) {
      finalView.push(
        <div key={i} className="multiple-choice-option">
          <label>
            <input
              name={questionCount}
              type="radio"
              onChange={() => handleChange(i)}
            />
            &nbsp;{choices[i]}
          </label>
        </div>
      );
    }

    return finalView;
  }

  getTrueFalse(questionCount, choices, handleChange) {
    let finalView = [];
    for (let i = 0; i < choices.length; i++) {
      finalView.push(
        <span key={i} className="true-false-option">
          <label>
            <input
              name={questionCount}
              type="radio"
              onChange={() => handleChange(choices[i])}
            />
            &nbsp;{choices[i]}
          </label>
        </span>
      );
    }

    return finalView;
  }

  getRange(questionCount, choices, handleChange) {
    let min = choices[0];
    let max = choices[1];
    return (
      <div className="range-question">
        <input
          id={questionCount}
          type="range"
          className="form-control-range"
          min={min}
          max={max}
          onChange={() => {
            const currentVal = document.getElementById(questionCount).value;
            this.getCurrentRangeVal(questionCount, currentVal);
            handleChange(currentVal);
          }}
        />
        <p>
          Value: <span id={'val' + questionCount} />
        </p>
      </div>
    );
  }

  getCurrentRangeVal(questionCount, currentVal) {
    document.getElementById('val' + questionCount).innerHTML = currentVal;
  }

  render() {
    return (
      <div>
        <div className="justify-content-center pad-bottom">
          <h3>Filters</h3>
        </div>
        {this.displayFilters()}
        <button
          className="filter-button"
          onClick={() => {
            this.props.fetchFilteredServiceProviders(this.newPredicates);
          }}
        >
          Filter
        </button>
      </div>
    );
  }
}

export default ServiceProviderFilter;
