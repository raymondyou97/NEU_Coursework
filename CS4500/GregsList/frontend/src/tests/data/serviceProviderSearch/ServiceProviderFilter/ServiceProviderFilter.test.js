'use es6';

import React from 'react';
import ServiceProviderFilter from '../../../../components/serviceProviderSearch/ServiceProviderFilter';
import { shallow } from 'enzyme';
import renderer from 'react-test-renderer';
import { configure } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';

import { SERVICE_QUESTION_LIST } from './ServiceProviderFilterData';

configure({ adapter: new Adapter() });

test('Snapshot renders with no service questions', () => {
  let serviceQuestions = [];
  const component = renderer.create(
    <ServiceProviderFilter serviceQuestions={serviceQuestions} />
  );

  let componentTree = component.toJSON();
  expect(componentTree).toMatchSnapshot();
});

test('Snapshot renders with service questions', () => {
  const component = renderer.create(
    <ServiceProviderFilter serviceQuestions={SERVICE_QUESTION_LIST} />
  );

  let componentTree = component.toJSON();
  expect(componentTree).toMatchSnapshot();
});

test('there exists 3 service questions', () => {
  let component = shallow(
    <ServiceProviderFilter serviceQuestions={SERVICE_QUESTION_LIST} />
  );

  expect(component.find('.service-filter-question').length).toEqual(3);
});

test('there exists 4 choices for the one multiple choice question', () => {
  let component = shallow(
    <ServiceProviderFilter serviceQuestions={SERVICE_QUESTION_LIST} />
  );

  expect(component.find('.multiple-choice-option').length).toEqual(4);
});

test('there exists 2 choices for the one true/false question', () => {
  let component = shallow(
    <ServiceProviderFilter serviceQuestions={SERVICE_QUESTION_LIST} />
  );

  expect(component.find('.true-false-option').length).toEqual(2);
});

test('there exists 1 range question', () => {
  let component = shallow(
    <ServiceProviderFilter serviceQuestions={SERVICE_QUESTION_LIST} />
  );

  expect(component.find('.range-question').length).toEqual(1);
});
