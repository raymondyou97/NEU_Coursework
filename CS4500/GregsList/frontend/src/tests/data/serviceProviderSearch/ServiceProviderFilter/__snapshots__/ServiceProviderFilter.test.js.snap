// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Snapshot renders with no service questions 1`] = `
<div>
  <div
    className="justify-content-center pad-bottom"
  >
    <h3>
      Filters
    </h3>
  </div>
</div>
`;

exports[`Snapshot renders with service questions 1`] = `
<div>
  <div
    className="justify-content-center pad-bottom"
  >
    <h3>
      Filters
    </h3>
  </div>
  <div
    className="service-filter-question"
  >
    <h5>
      Question 
      1
      : 
      question1
    </h5>
    <div
      className="multiple-choice-option"
    >
      <label>
        <input
          name={1}
          type="radio"
        />
         
        choice1
      </label>
    </div>
    <div
      className="multiple-choice-option"
    >
      <label>
        <input
          name={1}
          type="radio"
        />
         
        choice2
      </label>
    </div>
    <div
      className="multiple-choice-option"
    >
      <label>
        <input
          name={1}
          type="radio"
        />
         
        choice3
      </label>
    </div>
    <div
      className="multiple-choice-option"
    >
      <label>
        <input
          name={1}
          type="radio"
        />
         
        choice4
      </label>
    </div>
  </div>
  <div
    className="service-filter-question"
  >
    <h5>
      Question 
      2
      : 
      question2
    </h5>
    <span
      className="true-false-option"
    >
      <label>
        <input
          name={2}
          type="radio"
        />
         
        true
      </label>
    </span>
    <span
      className="true-false-option"
    >
      <label>
        <input
          name={2}
          type="radio"
        />
         
        false
      </label>
    </span>
  </div>
  <div
    className="service-filter-question"
  >
    <h5>
      Question 
      3
      : 
      question3
    </h5>
    <div
      className="range-question"
    >
      <input
        className="form-control-range"
        id={3}
        max="10"
        min="0"
        onChange={[Function]}
        type="range"
      />
      <p>
        Value: 
        <span
          id="val3"
        />
      </p>
    </div>
  </div>
</div>
`;
